Mathlib/Geometry/Manifold/Orientable.lean

```diff
@@ -0,0 +1,135 @@
+    OrientationReversing (g ∘ f) (u ∩ f ⁻¹' v) := by
+  intro x ⟨hxu, hxv⟩
+  rw [fderiv.comp x (hg.DifferentiableAt (f x) hxv) (hf.DifferentiableAt x hxu)]
+  unfold ContinuousLinearMap.det ContinuousLinearMap.comp
+  rw [(fderiv ℝ g (f x)).toLinearMap.det_comp (fderiv ℝ f x).toLinearMap]
+  exact mul_neg_of_neg_of_pos (hg (f x) hxv) (hf x hxu)
+
+lemma OrientationPreserving_comp_orientationReversing [Nontrivial H] [FiniteDimensional ℝ H]
+    {f g : H → H} {u v : Set H} (hf : OrientationReversing f u) (hg : OrientationPreserving g v) :
+    OrientationReversing (g ∘ f) (u ∩ f ⁻¹' v) := by
+  intro x ⟨hxu, hxv⟩
+  rw [fderiv.comp x (hg.DifferentiableAt (f x) hxv) (hf.DifferentiableAt x hxu)]
+  unfold ContinuousLinearMap.det ContinuousLinearMap.comp
+  rw [(fderiv ℝ g (f x)).toLinearMap.det_comp (fderiv ℝ f x).toLinearMap]
+  exact mul_neg_of_pos_of_neg (hg (f x) hxv) (hf x hxu)
+
+lemma orientationReversing_comp [Nontrivial H] [FiniteDimensional ℝ H] {f g : H → H} {u v : Set H}
+    (hf : OrientationReversing f u) (hg : OrientationReversing g v) :
+    OrientationPreserving (g ∘ f) (u ∩ f ⁻¹' v) := by
+  intro x ⟨hxu, hxv⟩
+  rw [fderiv.comp x (hg.DifferentiableAt (f x) hxv) (hf.DifferentiableAt x hxu)]
+  unfold ContinuousLinearMap.det ContinuousLinearMap.comp
+  rw [(fderiv ℝ g (f x)).toLinearMap.det_comp (fderiv ℝ f x).toLinearMap]
+  exact mul_pos_of_neg_of_neg (hg (f x) hxv) (hf x hxu)
+
+/-- The pregroupoid of orientation-preserving maps. -/
+def orientationPreservingPregroupoid [Nontrivial H] [FiniteDimensional ℝ H] : Pregroupoid H where
+  property f s := OrientationPreserving f s
+  comp hf hg _ _ _ _ hx := orientationPreserving_comp hf hg _ hx
+  id_mem := orientationPreserving_id _
+  locality {f u} _ h x hxu :=
+    have ⟨v, _, hxv, h⟩ := h x hxu
+    h x <| Set.mem_inter hxu hxv
+  congr {f g u} hu fg hf x hx := by
+    rw [(Filter.eventuallyEq_of_mem (hu.mem_nhds hx) fg).fderiv_eq]
+    exact hf x hx
+
+/-- The groupoid of orientation-preserving maps. -/
+def orientationPreservingGroupoid [Nontrivial H] [FiniteDimensional ℝ H] : StructureGroupoid H :=
+  orientationPreservingPregroupoid.groupoid
+
+end OrientationPreserving
+
+section OrientableManifold
+
+/-- Typeclass defining orientable manifolds. -/
+class OrientableManifold (H : Type*) [Nontrivial H] [NormedAddCommGroup H] [NormedSpace ℝ H]
+    [FiniteDimensional ℝ H] (M : Type*) [TopologicalSpace M] [ChartedSpace H M] extends
+  HasGroupoid M (@orientationPreservingGroupoid H _ _ _ _) : Prop
+
+/-- Typeclass defining orientable smooth manifolds. -/
+class OrientableSmoothManifold (H : Type*) [Nontrivial H] [NormedAddCommGroup H] [NormedSpace ℝ H]
+    [FiniteDimensional ℝ H] (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+    [OrientableManifold H M] (I : ModelWithCorners ℝ H M) [SmoothManifoldWithCorners I M] extends
+  SmoothManifoldWithCorners I M ∧ OrientableManifold H M : Prop
```

Comment:
Hm, I'd need to think if this does what we'd want: this definition asks for "M has a smooth atlas, and M has an orientable atlas" - whereas the obviously correct definition would be "M has a smooth orientable atlas" (i.e., one atlas which is both smooth and orientable).